### PR TITLE
Dualspace

### PR DESCRIPTION
**Pull request by nbouziani**
This PR mostly updates tests associated with the dualspace implementation. All the necessary block-level changes to have pyadjoint working with a Firedrake backend can be found in the corresponding PR:  https://github.com/firedrakeproject/firedrake/pull/2294

### CODE REVIEW

File: pyadjoint/reduced_functional.py [hunk 1]
```diff
@@ -124,13 +124,15 @@ def derivative(self, adj_input=1.0, options={}):
 
         # Scale adjoint input
         with stop_annotating():
-            adj_value = self.scale * adj_input
+            # Make sure adj_input is an OverloadedType
+            adj_input = create_overloaded_object(adj_input)
+            adj_input._ad_mul(self.scale)
```
Comment (dham) [inline pyadjoint/reduced_functional.py:129]: Split this to a separate PR.

**Review by dham — APPROVED** [review]

File: pyadjoint/reduced_functional.py (hunk 1 above)
Comment (nbouziani) [reply pyadjoint/reduced_functional.py:128]: Done, see https://github.com/dolfin-adjoint/pyadjoint/pull/112

File: .circleci/config.yml [hunk 2]
```diff
@@ -106,6 +106,10 @@ jobs:
       - run:
           name: install dependencies
           command: |
+            cd /home/firedrake/firedrake/src/ufl
+            git checkout fix_dualspace
+            cd /home/firedrake/firedrake/src/firedrake
+            git checkout dualspace_update
```
Comment (nbouziani) [inline .circleci/config.yml:112]:
```suggestion
```

**Review by dham — APPROVED** [review]
